src/catalog/dependency_manager.cpp

```diff
@@ -6,15 +6,236 @@
 #include "duckdb/main/client_context.hpp"
 #include "duckdb/main/database.hpp"
 #include "duckdb/parser/expression/constant_expression.hpp"
-#include "duckdb/catalog/mapping_value.hpp"
 #include "duckdb/catalog/dependency_list.hpp"
+#include "duckdb/common/enums/catalog_type.hpp"
+#include "duckdb/catalog/catalog_entry/dependency_catalog_entry.hpp"
+#include "duckdb/catalog/catalog_entry/duck_schema_entry.hpp"
+#include "duckdb/catalog/dependency_catalog_set.hpp"
 
 namespace duckdb {
 
-DependencyManager::DependencyManager(DuckCatalog &catalog) : catalog(catalog) {
+static void AssertMangledName(const string &mangled_name, idx_t expected_null_bytes) {
+#ifdef DEBUG
+	idx_t nullbyte_count = 0;
+	for (auto &ch : mangled_name) {
+		nullbyte_count += ch == '\0';
+	}
+	D_ASSERT(nullbyte_count == expected_null_bytes);
+#endif
+}
+
+MangledEntryName::MangledEntryName(const CatalogEntryInfo &info) {
+	static const auto NULL_BYTE = string(1, '\0');
+
+	auto &type = info.type;
+	auto &schema = info.schema;
+	auto &name = info.name;
+
+	this->name = CatalogTypeToString(type) + NULL_BYTE + schema + NULL_BYTE + name;
+	AssertMangledName(this->name, 2);
+}
+
+MangledDependencyName::MangledDependencyName(const MangledEntryName &from, const MangledEntryName &to) {
+	static const auto NULL_BYTE = string(1, '\0');
+	this->name = from.name + NULL_BYTE + to.name;
+	AssertMangledName(this->name, 5);
+}
+
+DependencyManager::DependencyManager(DuckCatalog &catalog)
+    : catalog(catalog), dependencies(catalog), dependents(catalog) {
+}
+
+string DependencyManager::GetSchema(CatalogEntry &entry) {
+	if (entry.type == CatalogType::SCHEMA_ENTRY) {
+		return entry.name;
+	}
+	return entry.ParentSchema().name;
+}
+
+MangledEntryName DependencyManager::MangleName(const CatalogEntryInfo &info) {
+	return MangledEntryName(info);
+}
+
+MangledEntryName DependencyManager::MangleName(CatalogEntry &entry) {
+	if (entry.type == CatalogType::DEPENDENCY_ENTRY) {
+		auto &dependency_entry = entry.Cast<DependencyCatalogEntry>();
+		return dependency_entry.MangledName();
+	}
+	auto type = entry.type;
+	auto schema = GetSchema(entry);
+	auto name = entry.name;
+	CatalogEntryInfo info {type, schema, name};
+
+	return MangleName(info);
+}
+
+DependencyInfo DependencyInfo::FromDependency(DependencyCatalogEntry &dep) {
```

Comment:
These should set the matching `from_type` or `to_type` for FromDependency/FromDependent